File: Two_Pointers/018.4Sum/018.4Sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> ret = new ArrayList<>();
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = nums.length - 1; j > 0; j--) {
                int left = i + 1;
                int right = j - 1;
                long sum = (long) target - nums[i] - nums[j];
                while (left < right) {
                    if ((long)nums[left] + nums[right] == sum) {
                        ret.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        left++;
                        right--;

                        while (left < right && nums[left] == nums[left - 1]){
                            left++;
                        }
                        while (left < right && nums[right] == nums[right + 1]) {
                            right--;
                        }
                    } else if ((long)nums[left] + nums[right] < sum) {
                        left++;
                    } else {
                        right--;
                    }
                }
                while (j - 1 > 0 && nums[j - 1] == nums[j]) {
                    j--;
                }
            }
            while (i + 1 < nums.length - 1 && nums[i + 1] == nums[i]) {
                i++;
            }
        }
        return ret;
    }
}

File: Two_Pointers/018.4Sum/018.4Sum.py
class Solution:
    def fourSum(self, nums:List[int], target: int) -> List[List[int]]:
        result = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i - 1] == nums[i]:
                continue

            for j in range(len(nums) - 1, 0, -1):
                if j + 1 < len(nums) and nums[j + 1] == nums[j]:
                    continue

                left = i + 1
                right = j - 1
                
                while left < right:
                    sum = nums[i] + nums[j] + nums[left] + nums[right]
                    if sum == target:
                        result.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1

                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1
                    elif sum < target:
                        left += 1
                    else:
                        right -= 1
        return result
                        
